File: test/tui/driver.py
# SPDX-License-Identifier: ISC
"""A pty + pyte harness for driving the clm TUI under test.

Spawns the real `clm` binary on a pseudo-terminal of a chosen size, feeds it
the mock server URL, and renders its output through a pyte terminal emulator so
tests can assert on the actual on-screen grid (text, cell attributes) and drive
keystrokes / resizes exactly as a user would.
"""
import fcntl
import os
import pty
import select
import signal
import struct
import termios
import time

import pyte

# Control bytes and xterm key sequences, for readable test scripts.
CTRL_A = b"\x01"
CTRL_B = b"\x02"
CTRL_E = b"\x05"
CTRL_F = b"\x06"
CTRL_K = b"\x0b"
CTRL_L = b"\x0c"
CTRL_U = b"\x15"
CTRL_Y = b"\x19"
ENTER = b"\r"
PGUP = b"\x1b[5~"
PGDN = b"\x1b[6~"
LEFT = b"\x1b[D"
RIGHT = b"\x1b[C"
# Cursor up/down in keypad application mode (SS3), which is what an app that
# called keypad(TRUE) puts the terminal into. Plain CSI (\x1b[A) is NOT what
# ncurses expects in that mode, so use these for arrow-key tests.
UP = b"\x1bOA"
DOWN = b"\x1bOB"


class Tui:
    def __init__(self, binary, url, rows=24, cols=80, extra_args=()):
        self.rows, self.cols = rows, cols
        self._screen = pyte.Screen(cols, rows)
        self._stream = pyte.ByteStream(self._screen)

        argv = [binary, "--url", url, *extra_args]
        self.pid, self._fd = pty.fork()
        if self.pid == 0:  # child
            os.environ["TERM"] = "xterm-256color"
            try:
                os.execv(argv[0], argv)
            finally:
                os._exit(127)
        self._set_winsize(rows, cols)
        self.pump(0.5)  # let it start and paint the first frame

    # ---- terminal plumbing ----
    def _set_winsize(self, rows, cols):
        fcntl.ioctl(self._fd, termios.TIOCSWINSZ,
                    struct.pack("HHHH", rows, cols, 0, 0))

    def pump(self, seconds):
        """Feed output into the emulator for a fixed window of time."""
        end = time.time() + seconds
        while time.time() < end:
            r, _, _ = select.select([self._fd], [], [], 0.05)
            if r:
                try:
                    data = os.read(self._fd, 65536)
                except OSError:
                    return
                if not data:
                    return
                self._stream.feed(data)

    def send(self, data):
        os.write(self._fd, data)

    def resize(self, rows, cols):
        self.rows, self.cols = rows, cols
        self._screen.resize(rows, cols)
        self._set_winsize(rows, cols)
        os.kill(self.pid, signal.SIGWINCH)
        self.pump(0.4)

    # ---- inspection ----
    def lines(self):
        return [self._screen.display[i].rstrip()
                for i in range(self.rows)]

    def text(self):
        return "\n".join(self.lines())

    def wait_for(self, needle, timeout=10.0):
        end = time.time() + timeout
        while time.time() < end:
            self.pump(0.2)
            if needle in self.text():
                return True
        return False

    def cell(self, y, x):
        return self._screen.buffer[y][x]

    def any_bold(self, needle):
        """True if every char of `needle` appears bold somewhere on screen."""
        found = ""
        for y in range(self.rows):
            row = self._screen.buffer[y]
            for x in range(self.cols):
                c = row[x]
                if c.bold and c.data and not c.data.isspace():
                    found += c.data
        return all(ch in found for ch in needle if not ch.isspace())

    def close(self):
        try:
            self.send(b"quit\r")
            self.pump(0.3)
        except OSError:
            pass
        try:
            os.kill(self.pid, signal.SIGTERM)
        except ProcessLookupError:
            pass
        try:
            os.close(self._fd)
        except OSError:
            pass
        try:
            os.waitpid(self.pid, 0)
        except ChildProcessError:
            pass

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

File: test/tui/driver.lua
-- SPDX-License-Identifier: ISC
--
-- A pty driver for the clm TUI under test. It starts the real binary on a
-- pseudo-terminal of a chosen size, points it at the mock server, and renders
-- its output through vt.lua, so a test can assert on the grid a user would
-- see and drive keystrokes and resizes the way a terminal does.

local sys = require("sys")
local vt = require("vt")

local M = {}

-- Shared session-state directory for every Tui in one run, so tests can read
-- the .jsonl session logs written under it.
M.STATE_HOME = sys.mkdtemp("/tmp/clm-tui-state-")

-- Control bytes and xterm key sequences, for readable test scripts.
M.CTRL_A = "\1"
M.CTRL_E = "\5"
M.CTRL_K = "\11"
M.CTRL_U = "\21"
M.CTRL_Y = "\25"
M.ESCAPE = "\27"
M.ENTER = "\r"
M.PGUP = "\27[5~"
M.PGDN = "\27[6~"
M.LEFT = "\27[D"
M.RIGHT = "\27[C"
-- Arrows in keypad application mode (SS3), which is what an application that
-- called keypad(TRUE) puts the terminal into. Plain CSI is not what ncurses
-- expects there.
M.UP = "\27OA"
M.DOWN = "\27OB"
M.HOME = "\27OH"
M.END = "\27OF"
-- Bracketed paste (mode 2004): what a terminal wraps pasted text in.
M.PASTE_START = "\27[200~"
M.PASTE_END = "\27[201~"

local Tui = {}
Tui.__index = Tui

-- binary is the clm under test, url the mock server. opts takes rows, cols,
-- args (extra argv) and env (extra environment).
function M.new(binary, url, opts)
	opts = opts or {}
	local rows = opts.rows or 24
	local cols = opts.cols or 80
	local argv = { binary, "--url", url }

	for _, a in ipairs(opts.args or {}) do
		argv[#argv + 1] = a
	end
	local env = {
		TERM = "xterm-256color",
		-- The transcript draws box characters, so the tui needs a
		-- utf-8 locale; a login shell that sets none would give it
		-- the C locale and single-byte output. macOS has no C.UTF-8.
		LC_ALL = sys.os == "darwin" and "en_US.UTF-8" or "C.UTF-8",
		-- The in-tree test config, so agent profiles resolve.
		XDG_CONFIG_HOME = M.TEST_DIR .. "/config",
		-- Session logs and scratch directories go to a per-run temp
		-- dir, never the developer's own state or cache.
		XDG_STATE_HOME = M.STATE_HOME,
		XDG_CACHE_HOME = M.STATE_HOME .. "/cache",
	}
	for k, v in pairs(opts.env or {}) do
		env[k] = v
	end

	local fd, pid = sys.spawn(argv, env, rows, cols)
	local t = setmetatable({
		fd = fd,
		pid = pid,
		rows = rows,
		cols = cols,
		screen = vt.new(rows, cols),
	}, Tui)
	-- The first curses frame arrives immediately. Drain it until the pty
	-- goes quiet rather than sleeping a fixed time per instance.
	t:drain(0.5, 0.03, true)
	return t
end

-- Feed output until the pty has been quiet for `quiet` seconds. `wait` lets
-- process startup take up to timeout; ordinary local operations return after
-- one quiet interval when no redraw was necessary.
function Tui:drain(timeout, quiet, wait)
	timeout = timeout or 0.5
	quiet = quiet or 0.03
	local deadline = sys.now() + timeout
	local received = false

	while sys.now() < deadline do
		local left = deadline - sys.now()
		local interval = (received or not wait) and quiet or left

		if interval > left then
			interval = left
		end
		local data = sys.read(self.fd, math.floor(interval * 1000))
		if data == nil or data == false then
			return
		end
		received = true
		self.screen:feed(data)
	end
end

-- Feed output into the emulator for a fixed window of time.
function Tui:pump(seconds)
	local deadline = sys.now() + seconds

	while sys.now() < deadline do
		local data = sys.read(self.fd, 50)
		if data == false then
			return
		end
		if data ~= nil then
			self.screen:feed(data)
		end
	end
end

function Tui:send(data)
	sys.write(self.fd, data)
end

function Tui:resize(rows, cols)
	self.rows, self.cols = rows, cols
	self.screen:resize(rows, cols)
	sys.winsize(self.fd, rows, cols)
	sys.kill(self.pid, sys.SIGWINCH)
	self:drain()
end

-- Screen rows, trailing blanks removed, one-based.
function Tui:lines()
	local out = {}

	for y = 0, self.rows - 1 do
		out[y + 1] = (self.screen:line(y):gsub("%s+$", ""))
	end
	return out
end

function Tui:text()
	return table.concat(self:lines(), "\n")
end

function Tui:wait_for(needle, timeout)
	local deadline = sys.now() + (timeout or 10.0)

	while sys.now() < deadline do
		self:pump(0.2)
		if self:text():find(needle, 1, true) then
			return true
		end
	end
	return false
end

-- The cell at row y, column x, both zero-based.
function Tui:cell(y, x)
	return self.screen:cell(y, x)
end

-- Where the terminal would draw the caret: row, column, both zero-based.
function Tui:cursor()
	return self.screen:cursor()
end

-- True when every character of needle appears bold somewhere on screen.
function Tui:any_bold(needle)
	local found = {}

	for y = 0, self.rows - 1 do
		for x = 0, self.cols - 1 do
			local c = self.screen:cell(y, x)
			if (c.attr & vt.BOLD) ~= 0 and c.ch:match("%S") then
				found[c.ch] = true
			end
		end
	end
	for ch in needle:gmatch("%S") do
		if not found[ch] then
			return false
		end
	end
	return true
end

function Tui:close()
	if self.pid ~= nil then
		self:send("quit\r")
		-- A normal quit exits promptly. Wait only briefly, then fall
		-- back to a signal for a wedged client.
		local deadline = sys.now() + 0.3
		while sys.now() < deadline do
			if sys.wait(self.pid, true) == self.pid then
				self.pid = nil
				break
			end
			self:drain(0.05, 0.01)
		end
	end
	if self.pid ~= nil then
		sys.kill(self.pid, sys.SIGTERM)
	end
	sys.close(self.fd)
	if self.pid ~= nil then
		sys.wait(self.pid, false)
		self.pid = nil
	end
end

return M
